### ISSUE TITLE
Change detection algorithm doesn't add the features to the Added layer the first time

### ISSUE THREAD

**Issue by antoniolocandro** [qgis/QGIS#46271]
### What is the bug or the crash?

When the algorithm is run the first time it displays that there are features added but they are not really added to the layer, it takes a second run of the algorithm to get the correct results

First run parameters
![image](https://user-images.githubusercontent.com/4627073/143937565-4e1a8450-890f-446a-b61d-1820d12dd2d7.png)

Video Explaining the process
https://user-images.githubusercontent.com/4627073/143937476-2be71622-db40-4800-a27d-eb30d855e91b.mp4

### Steps to reproduce the issue

Attached is the sample data, however the original files are in gml format so I do not know if that makes a difference. I tried with a different profile with the gml I had issues but with the esri shapefiles it worked OK

### Versions

QGIS version
3.22.1-Białowieża
QGIS code revision
663dcf8fb9
Qt version
5.15.2
Python version
3.9.5
GDAL/OGR version
3.4.0
PROJ version
8.2.0
EPSG Registry database version
v10.038 (2021-10-21)
GEOS version
3.10.0-CAPI-1.16.0
SQLite version
3.35.2
PDAL version
2.3.0
PostgreSQL client version
13.0
SpatiaLite version
5.0.1
QWT version
6.1.3
QScintilla2 version
2.11.5
OS version
Windows 10 Version 2009

Active Python plugins
BivariateLegend
1.0.2
BivariateRenderer
0.3
cartography_tools
1.2.1
DigitizingTools
1.5.1
GlobeBuilder
v0.6.0
ImageServerConnector
2.1.1
LAStools
1.4
latlontools
3.5.9
MemoryLayerSaver
4.0.4
mmqgis
2021.9.10
ORStools
1.4.0
qconsolidate3
0.2.0
QNEAT3
1.0.4
qpansops
0.1
quick_map_services
0.19.26
selectThemes
3.0.1
SRTM-Downloader
3.1.15
db_manager
0.1.20
grassprovider
2.12.99
otbprovider
2.12.99
processing
2.12.99
sagaprovider
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

I also tried with different profiles, only difference seems to be the GML structure.

**Comment by nyalldawson** [comment]
Can you supply the datasets?

**Comment by antoniolocandro** [comment]
[test_data.zip](https://github.com/qgis/QGIS/files/7621367/test_data.zip)
Forgot! These as shapefiles

The original XML ones I can't provide publicly as it contains other data also

**Comment by gioman** [comment]
Cannot replicate with the provided dataset.

**Comment by antoniolocandro** [comment]
Here is the original data, just drag and drop each xml file, the gfs is already created.

Then rename one of the routes new the other old to keep tabs on the algorithm checkbox and run, should read 27 unchanged 9 deleted 9 added and the first time the 9 added are not shown. Use Exact Match no attributes

[aixm.zip](https://github.com/qgis/QGIS/files/7636560/aixm.zip)

**Comment by gioman** [comment]
> should read 27 unchanged 9 deleted 9 added

@antoniolocandro this is the log on the first run


```
27 features unchanged
9 features added
36 features deleted
```

**Comment by antoniolocandro** [comment]
This is my setup
![image](https://user-images.githubusercontent.com/4627073/144305615-c9f4f6e4-3767-49f7-978a-a964e09542bb.png)
![image](https://user-images.githubusercontent.com/4627073/144305638-b8e4b31e-c089-4bc1-ba3c-46cb4e824d38.png)

This is what I get

**Comment by antoniolocandro** [comment]
If I run it the other way around with layers I get your values but then the layer is not created correctly, noticed unchanged has 0 features instead of 27 reported in the log

![image](https://user-images.githubusercontent.com/4627073/144305988-98a979a8-88cd-49c4-8e2b-51dda0638886.png)

**Comment by gioman** [comment]
> This is my setup

@antoniolocandro I used the "route segments" layers in your XMLs.

**Comment by antoniolocandro** [comment]
Yes that is correct, so the thing is when running the algorithm the log detects correctly the values but the actual layer the first time it  creates it doesn't display the features. Like they aren't written. I am using temporary scratch layers also

**Comment by gioman** [comment]
> it doesn't display the features

it is fine here, all the output layers have features and attributes.
